packages/ketcher-react/src/script/editor/tool/templatePreview.ts

```diff
@@ -0,0 +1,233 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ ***************************************************************************/
+
+import {
+  Vec2,
+  fromItemsFuse,
+  fromTemplateOnAtom,
+  fromTemplateOnBondAction,
+  getItemsToFuse,
+  Action,
+  fromTemplateOnCanvas,
+  fromMultipleMove,
+  getHoverToFuse,
+} from 'ketcher-core';
+import Editor from '../Editor';
+import { MODES } from 'src/constants';
+import { getAngleFromEvent, getBondFlipSign } from './template';
+
+const PREVIEW_DELAY = 300;
+type ClosestItemType = { map: string; id: number; dist: number };
+
+function getUniqueCiId(ci: ClosestItemType) {
+  return `${ci.id}-${ci.map}`;
+}
+
+class TemplatePreview {
+  private readonly editor: Editor;
+  private readonly template;
+  private readonly mode: 'fg' | null;
+  private floatingPreviewAction: Action | null;
+  private connectedPreviewAction: Action | null;
+  private connectedPreviewTimeout: ReturnType<typeof setTimeout> | null = null;
+  private lastPreviewId: string | null;
+
+  private floatingPreview: {
+    atoms: number[];
+    bonds: number[];
+  } | null;
+
+  private position: Vec2;
+  private previousPosition: Vec2;
+
+  constructor(editor: Editor, template, mode) {
+    this.editor = editor;
+    this.template = template;
+    this.floatingPreviewAction = new Action();
+    this.connectedPreviewAction = null;
+    this.connectedPreviewTimeout = null;
+    this.lastPreviewId = null;
+    this.mode = mode;
+    this.floatingPreview = null;
+    this.position = new Vec2();
+    this.previousPosition = new Vec2();
+  }
+
+  private get struct() {
+    return this.restruct.molecule;
+  }
+
+  private get restruct() {
+    return this.editor.render.ctab;
+  }
+
+  private get isModeFunctionalGroup(): boolean {
+    return this.mode === MODES.FG;
+  }
+
+  hidePreview() {
+    this.hideConnectedPreview();
+    this.hideFloatingPreview();
+  }
+
+  private getPreviewTarget() {
+    const ci: ClosestItemType | null = this.editor.findItem(event, [
+      'atoms',
+      'bonds',
+    ]);
+
+    if (ci && this.restruct.molecule[ci.map].get(ci.id)?.isPreview === false) {
+      return ci;
+    } else {
+      return null;
+    }
+  }
+
+  movePreview(event: PointerEvent) {
+    this.position = this.editor.render.page2obj(event);
+
+    const ci = this.getPreviewTarget();
+    const isMouseAwayFromAtomsAndBonds = !ci;
+    const isPreviewTargetChanged =
+      ci && this.lastPreviewId !== getUniqueCiId(ci);
+
+    const shouldHidePreview =
+      isMouseAwayFromAtomsAndBonds || isPreviewTargetChanged;
+
+    const shouldShowPreview =
+      ci && !this.connectedPreviewAction && !this.connectedPreviewTimeout;
```

Comment:
I would propose to rename it to shouldHideConnectedPreview and shouldShowConnectedPreview